src/Features/CSharp/Portable/UpgradeProject/CSharpAsyncMainDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,130 @@
+                DiagnosticSeverity.Warning,
+                isEnabledByDefault: true);
+
+        private static readonly ImmutableDictionary<string, string> s_diagnosticProperties =
+            ImmutableDictionary.Create<string, string>().Add(DiagnosticPropertyConstants.RequiredLanguageVersion, "7.1");
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } =
+            ImmutableArray.Create(s_diagnosticDescriptor);
+
+        public DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        public bool OpenFileOnly(Workspace workspace) => false;
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(CompilationStartAction);
+        }
+
+        private static void CompilationStartAction(CompilationStartAnalysisContext context)
+        {
+            var compilation = context.Compilation;
+            if (compilation.Options.OutputKind != OutputKind.ConsoleApplication &&
+                compilation.Options.OutputKind != OutputKind.WindowsApplication &&
+                compilation.Options.OutputKind != OutputKind.WindowsRuntimeApplication)
+            {
+                return;
+            }
+
+            var taskType = compilation.GetTypeByMetadataName("System.Threading.Tasks.Task");
+            if (taskType == null)
+            {
+                return;
+            }
+
+            var taskOfTType = compilation.GetTypeByMetadataName("System.Threading.Tasks.Task`1");
+            if (taskOfTType == null)
+            {
+                return;
+            }
+
+            var types = new AnalysisTypes(
+                taskType,
+                taskOfTType.Construct(compilation.GetSpecialType(SpecialType.System_Int32)),
+                compilation.CreateArrayTypeSymbol(compilation.GetSpecialType(SpecialType.System_String)));
+
+            context.RegisterSymbolAction(symbolContext => AnalyzeSymbol(symbolContext, types), SymbolKind.Method);
+        }
+
+        private static void AnalyzeSymbol(SymbolAnalysisContext context, AnalysisTypes types)
+        {
+            var method = (IMethodSymbol)context.Symbol;
+
+            if (IsValidMainMethodKind(method) &&
+                IsValidAsyncMainParameterList(method.Parameters, types) &&
+                IsValidAsyncMainReturnType(method.ReturnType, types) &&
+                method.DeclaringSyntaxReferences.Length == 1)
+            {
+                var declarationReference = method.DeclaringSyntaxReferences[0];
+
+                var parseOptions = (CSharpParseOptions)declarationReference.SyntaxTree.Options;
+                if (parseOptions.Kind == SourceCodeKind.Regular &&
+                    parseOptions.LanguageVersion < LanguageVersion.CSharp7_1)
+                {
+                    var entryPoint = context.Compilation.GetEntryPoint(context.CancellationToken);
```

Comment:
@CyrusNajmabadi  I think I remember you mentioning somewhere that this is really expensive. Is that the case? If so, should I do this in some other place such as compilation end action? Or is it so bad that we can't use it at all?